File: problem_079.py
#!/usr/bin/python

print("Project Euler: 79")
print("Given that the three characters are always asked for in order, analyse the file so as to determine the shortest possible secret passcode of unknown length.\n")

Before1 = []
Before2 = []
Before3 = []
Before6 = []
Before7 = []
Before8 = []
Before9 = []
Before0 = []

ReadFile = open("problem_079.txt")
while 1:
    Line = ReadFile.readline()
    if not Line: break
    Line = Line[0:3]
    #print Line

    Digit1 = Line[0]
    Digit2 = Line[1]
    Digit3 = Line[2]

    if   Digit2 == "1": Before1.append(Digit1)
    elif Digit2 == "2": Before2.append(Digit1)
    elif Digit2 == "3": Before3.append(Digit1)
    elif Digit2 == "6": Before6.append(Digit1)
    elif Digit2 == "7": Before7.append(Digit1)
    elif Digit2 == "8": Before8.append(Digit1)
    elif Digit2 == "9": Before9.append(Digit1)
    elif Digit2 == "0": Before0.append(Digit1)

    if   Digit3 == "1": Before1.append(Digit2)
    elif Digit3 == "2": Before2.append(Digit2)
    elif Digit3 == "3": Before3.append(Digit2)
    elif Digit3 == "6": Before6.append(Digit2)
    elif Digit3 == "7": Before7.append(Digit2)
    elif Digit3 == "8": Before8.append(Digit2)
    elif Digit3 == "9": Before9.append(Digit2)
    elif Digit3 == "0": Before0.append(Digit2)

# Make lists unique
Before1 = list(set(Before1))
Before2 = list(set(Before2))
Before3 = list(set(Before3))
Before6 = list(set(Before6))
Before7 = list(set(Before7))
Before8 = list(set(Before8))
Before9 = list(set(Before9))
Before0 = list(set(Before0))

Answer = []
while 1:
    for i in 1, 2, 3, 6, 7, 8, 9, 0:
        if len(eval("Before"+str(i)))==0:
            if not i in Answer:
                #print i
                Answer.append(i)
                break
        for n in range(len(Answer)):
            if str(Answer[n]) in eval("Before"+str(i)):
                PopMe=eval("Before"+str(i)).index(str(Answer[n]))
                eval("Before"+str(i)).pop(PopMe)
    if len(Answer) == 8:
        break

print Answer

File: problem_079.rs
// -*- compile-command: "rustc -o problem_079_rs problem_079.rs" -*-
// Copyright (c) 2016 Michael Caldwell
use std::fs::File;
use std::io::Read;

fn is_present(a: [i32; 3], n: i32) -> bool {
    for i in 0..3 {
        if a[i] == n {
            return true;
        }
    }
    return false;
}

fn is_rightmost(a: [i32; 3], n: i32) -> bool {
    // See if number is in this set
    let mut seen = false;
    for i in 0..3 {
        if a[i] == n {
            seen = true;
        }
    }
    if seen == false { return true; }

    for i in (0..3).rev() {
        if a[i] == -1 { continue; }
        if a[i] == n { return true; }
        else { return false; }
    }
    return true;
}

fn main() {
    println!("Project Euler - Problem 79:");
    println!("Given that the three characters are always asked for in order, analyse the file so as to determine the shortest possible secret passcode of unknown length.\n");

    // Import data from file
    let mut data = String::new();
    let mut f = File::open("problem_079.txt").unwrap();
    f.read_to_string(&mut data).unwrap();

    // Split codes into vector
    let mut codes_ascii: Vec<&str> = data.split('\n').collect();

    // Sort codes alphabetically
    codes_ascii.sort();

    // Convert codes from ASCII Text to Integers (new vector)
    let mut codes_int = Vec::new();
    for code in codes_ascii {
        let tmp = code.parse::<usize>();
        if tmp.is_err() == false {
            let mut int = tmp.unwrap();
            let mut array: [i32; 3] = [0, 0, 0];
            for i in (0..3).rev() {
                array[i] = (int % 10) as i32;
                int /= 10;
            }
            codes_int.push(array);
        }
    }

    // search for each number
    let mut answer = Vec::new();
    loop {
        for i in 0..10 {
            // See if number is present
            let mut present = false;
            for n in 0..codes_int.len(){
                if is_present(codes_int[n], i as i32) == true {
                    present = true;
                    break;
                }
            }
            if present == false { continue; }

            // See if number is right most
            let mut right = true;
            for n in 0..codes_int.len(){
                if is_rightmost(codes_int[n], i as i32) == false {
                    right = false;
                    break;
                }
            }

            if right == true {
                answer.push(i);

                // Mark code number as -1 to indicate used
                for n in 0..codes_int.len(){
                    for j in 0..3 {
                        if codes_int[n][j] == i as i32{
                            codes_int[n][j] = -1;
                        }
                    }
                }

                break;
            }
        }

        // See if we are done by checking if all numbers are used (-1)
        let mut done = true;
        for n in 0..codes_int.len(){
            for j in 0..3 {
                if codes_int[n][j] > -1 {
                    done = false;
                }
            }
        }

        if done == true {
            break;
        }
    }

    // Print answer
    for i in (0..answer.len()).rev() {
        print!("{}", answer[i]);
    }
    println!("");
}
